File: src/main/java/net/frozenblock/configurableeverything/entity/util/EntityAttributeAmplifier.java
package net.frozenblock.configurableeverything.entity.util;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import java.util.List;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.ResourceKey;
import net.minecraft.world.entity.EntityType;

public record EntityAttributeAmplifier(ResourceKey<EntityType<?>> entity, String entityName, List<AttributeAmplifier> amplifiers) {

	public static final Codec<EntityAttributeAmplifier> CODEC = RecordCodecBuilder.create(instance ->
		instance.group(
			ResourceKey.codec(Registries.ENTITY_TYPE).fieldOf("entity").forGetter(EntityAttributeAmplifier::entity),
			Codec.STRING.fieldOf("entityName").forGetter(EntityAttributeAmplifier::entityName),
			AttributeAmplifier.CODEC.listOf().fieldOf("amplifiers").forGetter(EntityAttributeAmplifier::amplifiers)
		).apply(instance, EntityAttributeAmplifier::new)
	);
}

File: src/main/java/net/frozenblock/configurableeverything/entity/util/EntityAttributeAmplifier.kt
package net.frozenblock.configurableeverything.entity.util

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.registries.Registries
import net.minecraft.resources.ResourceKey
import net.minecraft.world.entity.EntityType

data class EntityAttributeAmplifier(
	@JvmField var entity: ResourceKey<EntityType<*>>,
	@JvmField var entityName: String,
	@JvmField var amplifiers: MutableList<AttributeAmplifier>
) {
    companion object {
        @JvmField
		val CODEC: Codec<EntityAttributeAmplifier> = RecordCodecBuilder.create { instance ->
            instance.group(
                ResourceKey.codec(Registries.ENTITY_TYPE).fieldOf("entity").forGetter(EntityAttributeAmplifier::entity),
                Codec.STRING.fieldOf("entityName").forGetter(EntityAttributeAmplifier::entityName),
                AttributeAmplifier.CODEC.mutListOf().fieldOf("amplifiers").forGetter(EntityAttributeAmplifier::amplifiers)
            ).apply(instance, ::EntityAttributeAmplifier)
        }
    }
}
